collect_app/src/main/java/org/odk/collect/android/tasks/ServerPollingJob.java

```diff
@@ -0,0 +1,204 @@
+
+    @Override
+    @NonNull
+    protected Result onRunJob(@NonNull Params params) {
+        if (!isDeviceOnline()) {
+            GeneralSharedPreferences.getInstance().save(POLL_SERVER_IMMEDIATELY_AFTER_RECEIVING_NETWORK, true);
+            return Result.FAILURE;
+        } else {
+            GeneralSharedPreferences.getInstance().reset(POLL_SERVER_IMMEDIATELY_AFTER_RECEIVING_NETWORK);
+            HashMap<String, FormDetails> formList = DownloadFormListUtils.downloadFormList(true);
+
+            if (formList != null && !formList.containsKey(DL_ERROR_MSG)) {
+                if (formList.containsKey(DL_AUTH_REQUIRED)) {
+                    AuthDialogUtility.setWebCredentialsFromPreferences();
+                    formList = DownloadFormListUtils.downloadFormList(true);
+
+                    if (formList == null || formList.containsKey(DL_AUTH_REQUIRED) || formList.containsKey(DL_ERROR_MSG)) {
+                        return Result.FAILURE;
+                    }
+                }
+
+                List<FormDetails> newDetectedForms = new ArrayList<>();
+                for (FormDetails formDetails : formList.values()) {
+                    if (formDetails.isNewerFormVersionAvailable() || formDetails.areNewerMediaFilesAvailable()) {
+                        newDetectedForms.add(formDetails);
+                    }
+                }
+
+                if (!newDetectedForms.isEmpty()) {
+                    if (GeneralSharedPreferences.getInstance().getBoolean(KEY_AUTOMATIC_UPDATE, false)) {
+                        final HashMap<FormDetails, String> result = new FormDownloader().downloadForms(newDetectedForms);
+                        informAboutNewDownloadedForms(Collect.getInstance().getString(R.string.forms_downloaded) + "\n\n" + FormDownloadList.getDownloadResultMessage(result));
+                    } else {
+                        for (FormDetails formDetails : newDetectedForms) {
+                            String manifestFileHash = formDetails.getManifestFileHash() != null ? formDetails.getManifestFileHash() : "";
+                            String formVersionHash = FormDownloader.getMd5Hash(formDetails.getHash()) + manifestFileHash;
+                            if (!wasThisNewerFormVersionAlreadyDetected(formVersionHash)) {
+                                updateLastDetectedFormVersionHash(formDetails.getFormID(), formVersionHash);
+                            } else {
+                                newDetectedForms.remove(formDetails);
+                            }
+                        }
+
+                        if (!newDetectedForms.isEmpty()) {
+                            informAboutNewAvailableForms();
+                        }
+                    }
+                }
+                return Result.SUCCESS;
+            } else {
+                return Result.FAILURE;
+            }
+        }
+    }
+
+    public static void schedulePeriodicJob(String selectedOption) {
```

Comment:
It may be worth considering extracting out this method to a another class, it doesn't strictly belong in a  ServerPollingJob (In the interests of Single Responsiblity :-) ).

Reply:
In the example here https://github.com/evernote/android-job it's in the same class and I also think it's not a problem.